File: chapter_6/table_printer.py
import time


def printTable(tableData):
    # Step 1: Determine the maximum width of each column
    colWidths = [0] * len(tableData)

    for i in range(len(tableData)):
        colWidths[i] = max(len(item) for item in tableData[i])

    # Step 2: Print the table row by row
    numRows = len(tableData[0])
    for row in range(numRows):
        for col in range(len(tableData)):
            print(tableData[col][row].rjust(colWidths[col]), end=" ")
        print()  # Newline after each row


# Example usage
start_time = time.time()
tableData = [
    ["apples", "oranges", "cherries", "banana"],
    ["Alice", "Bob", "Carol", "David"],
    ["dogs", "cats", "moose", "goose"],
]

printTable(tableData)
end_time = time.time()
elapsed_time = end_time - start_time

print(f"Time taken: {elapsed_time:.6f} seconds")

File: chapter_6/src/bin/table_printer.rs
use std::time::Instant;

fn print_table(table_data: [[&str; 4]; 3]) {
    let mut col_widths = vec![0; 4];
    // loop over the table_data to get the max width of every col
    for (i, inner_list) in table_data.iter().enumerate() {
        col_widths[i] = inner_list.iter().map(|s| s.len()).max().unwrap();
    }
    // println!("{:?}", col_width);
    // now we will loop over the rows and print each line
    let num_rows = table_data[0].len();
    for row in 0..num_rows {
        for col in 0..table_data.len() {
            print!("{:>width$} ", table_data[col][row], width = col_widths[col]);
        }
        println!("")
    }
}

fn main() {
    let start_time = Instant::now();
    const TABLE_DATA: [[&str; 4]; 3] = [
        ["apples", "oranges", "cherries", "banana"],
        ["Alice", "Bob", "Carol", "David"],
        ["dogs", "cats", "moose", "goose"],
    ];
    print_table(TABLE_DATA);

    let elapsed_time = start_time.elapsed();
    println!("Time taken: {:.6} seconds", elapsed_time.as_secs_f64());
}
